### ISSUE TITLE
Move from redis to psql as a data store

### ISSUE THREAD

**Issue by TomasTomecek** [packit-service/packit-service#459]
# Plan

## Stage 1
* [ ] store data in psql and redis in parallel, read data from psql, fall back to redis
  * [x] **required** change REST API code to do this [p-s/service/api](https://github.com/packit-service/packit-service/tree/master/packit_service/service/api) #479
    * assigned to @IceWreck 
  * [x] **required** turn whitelist into psql (schema needs to be created) [p-s/worker/whitelist](https://github.com/packit-service/packit-service/blob/master/packit_service/worker/whitelist.py)
    * assigned to @IceWreck 
  * [ ] store installations-related data into psql (also needs schema and design) [p-s/worker/handlers/github](https://github.com/packit-service/packit-service/blob/master/packit_service/worker/handlers/github_handlers.py)
    * assigned to @rishavanand
  * [x] store testing farm data in psql [p-s/worker/handlers/testing-farm](https://github.com/packit-service/packit-service/blob/master/packit_service/worker/handlers/testing_farm_handlers.py) (done by @lachmanfrantisek in #502 )
  * [x] store information about all tasks/jobs executed in p-s into psql (needs schema and design)
    (done by @lbarcziova in #520)
  * [ ] make sure that data we store in redis are a **subset** of what we put in psql
    * assigned to @rishavanand
  * [ ] get a bottle of champagne and prepare for p2
    * assigned to @rishavanand

## Stage 2
* [ ] create a script to read data from redis and transform them into sqlalchemy objects and save them into psql
  * make the script idempotent - can be run multiple times on the same data set without creating duplicate entries
  * remove the data from redis
  * DON'T RUN THE SCRIPT IN PROD JUST YET

## Stage 3
* [ ] once everything is migrated into psql, get rid off all the redis-related code
* [ ] deploy the code to prod
* [ ] run the script from Stage 2 on prod

---

* needs-design mean that you should first propose a solution (mainly the schema) to the maintainers and get it ack'd before working on the code itself

please update this as needed

**Comment by IceWreck** [comment]
Hey, I can work on atleast parts of Stage 1, though after 9th March (the first two checkboxes)

**Comment by grapheo12** [comment]
I am also interested in this. Can we work together? @TomasTomecek  @IceWreck

**Comment by grapheo12** [comment]
@TomasTomecek I have an action plan in mind for this:

1. Let's use the raw SQLAlchemy (not the flask_sqlalchemy plugin) so that the db connection object can be used outside the context of the Flask app.

2. Let's take the code for models out of the service folder as the definitions might be useful for other programs in workers and handlers too. (Possibly will not let some complicated import issues later on)

3. In models.py we can let the `Model` class inherit the Model class exposed by SQLAlchemy. Let's adhere to strict duck-typing so that all the methods signatures are preserved. But under the hood, let's use psql first and then redis as a fallback.

4. Later this redis part can be dropped at will.

Please review if this falls in line with what is wanted.

**Comment by IceWreck** [comment]
> I am also interested in this. Can we work together? @TomasTomecek @IceWreck

Yes, sure.

**Comment by IceWreck** [comment]
@grapheo12

**Comment by grapheo12** [comment]
> The postgres models.py is already outside the service folder
Thanks for pointing out. I probably missed that out.
Is there any reason to keep them separate? We can possibly merge them together.

**Comment by IceWreck** [comment]
> > The postgres models.py is already outside the service folder
> 
> Thanks for pointing out. I probably missed that out.
> Is there any reason to keep them separate? We can possibly merge them together.

The one inside the service folder is the redis models.py and the one outside is the postgres models.py

**Comment by grapheo12** [comment]
Yeah but class definitions can be made such that same model can be used for both. That'd be better since then we will not have to modify each and every use of the models with two method calls.

**Comment by IceWreck** [comment]
> Yeah but class definitions can be made such that same model can be used for both. That'd be better since then we will not have to modify each and every use of the models with two method calls.



Redis isn't a good way to store large amounts of data, but it is what was being used before. It is slowly being replaced postgres (some of the code migration is already done).
@TomasTomecek wants to replace the leftover redis code (such as in the Rest API) with postgres equivalent code. @grapheo12 atleast read the issue post dude.

**Comment by grapheo12** [comment]
> @grapheo12 Read the issue post.

I am talking of Stage 1 specifically.

**Comment by IceWreck** [comment]
> I am talking of Stage 1 specifically.

In multiple places throughout this repo, data is stored in the redis models.py We have to change it to use the models in postgres models.py
In some cases equivalent structures have already been made in the postgres models.py In other cases we will have to modify the postgres models.py to add the required schema

**Comment by TomasTomecek** [comment]
@IceWreck thanks for the explanation, you're completely right! I assigned the first two points from p1 to you - good luck!

@grapheo12 if you're still interested in working on this, please let us know which point(s) from the main listing; you have raised some good points in your post, though that's about refactoring and improving how p-s deals with sqlalch - please create a new issue for that, since that's out of scope of this one - with this issue, we're trying to shred packit-service/service/models.py as @IceWreck pointed out

**Comment by grapheo12** [comment]
@TomasTomecek I would like to work on 3 and 4.

**Comment by rishavanand** [comment]
@TomasTomecek I would like to take up 5, 6 and 7 (of course) of stage 1.

**Comment by TomasTomecek** [comment]
@grapheo12 @rishavanand thank you and best of luck to you! please reach out to us if you need any help here

**Comment by TomasTomecek** [comment]
@rishavanand @grapheo12 how are you guys doing, any progress?

**Comment by rishavanand** [comment]
@TomasTomecek I was occupied with my GSoC proposal. I've also included this issue in the proposal but I'll be starting working on it this week itself.

**Comment by grapheo12** [comment]
@TomasTomecek Sorry. I got occupied in some other work. I will not be applying for GSoC this year, most probably. Will look into packit in my free time. Can you please reassign someone?

**Comment by rishavanand** [comment]
@TomasTomecek I'll be starting tomorrow with this issue, you may re-assign the tasks to me.

**Comment by TomasTomecek** [comment]
> @TomasTomecek I was occupied with my GSoC proposal. I've also included this issue in the proposal but I'll be starting working on it this week itself after resolving [this dashboard issue](https://github.com/packit-service/dashboard/issues/29)

Okay, got it!

> @TomasTomecek Sorry. I got occupied in some other work. I will not be applying for GSoC this year, most probably. Will look into packit in my free time. Can you please reassign someone?

Thanks for the update, best of luck!

**Comment by lbarcziova** [comment]
@rishavanand I have started to write the script to migrate data from redis to psql and it requires the model for installations in psql, would you like to work on this part too? (I can do it if not)

**Comment by rishavanand** [comment]
@lbarcziova Yeah I can take it up, instead of starting with task 5 I'll start with task 3. I might be a bit slow, though.

**Comment by lbarcziova** [comment]
hey @rishavanand,  how is it going?

**Comment by rishavanand** [comment]
hey @lbarcziova I'm still stuck at local setup

**Comment by lbarcziova** [comment]
oh, so then if you do not mind, I will do the installation model by myself so that I can finish the migration script

**Comment by rishavanand** [comment]
sure, go ahead

**Comment by lbarcziova** [comment]
This is already done.
